src/include/network/peloton_rpc_handler_task.h

```diff
@@ -13,22 +13,135 @@
+    LOG_DEBUG("Database oid: %d", database_oid);
+    LOG_DEBUG("Index oid: %d", index_oid);
+
+    auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    // Drop index. Fail if it doesn't exist.
+    auto catalog = catalog::Catalog::GetInstance();
+    try {
+      catalog->DropIndex(database_oid, index_oid, txn);
+    } catch (CatalogException e) {
+      LOG_ERROR("Drop Index Failed");
+      txn_manager.AbortTransaction(txn);
+      return kj::NEVER_DONE;
+    }
+    txn_manager.CommitTransaction(txn);
     return kj::READY_NOW;
   }
-};
 
+  kj::Promise<void> createIndex(CreateIndexContext request) override {
+    LOG_DEBUG("Received RPC to create index");
+    auto database_oid = request.getParams().getRequest().getDatabaseOid();
+    auto table_oid = request.getParams().getRequest().getTableOid();
+    auto col_oids = request.getParams().getRequest().getKeyAttrOids();
+    auto is_unique = request.getParams().getRequest().getUniqueKeys();
+    auto index_name = request.getParams().getRequest().getIndexName();
+    std::vector<oid_t> col_oid_vector;
+    LOG_DEBUG("Database oid: %d", database_oid);
+    LOG_DEBUG("Table oid: %d", table_oid);
+    for (auto col : col_oids) {
+      LOG_DEBUG("Col oid: %d", col);
+      col_oid_vector.push_back(col);
+    }
+
+    auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+    auto txn = txn_manager.BeginTransaction();
+
+    // Create index. Fail if it already exists.
+    auto catalog = catalog::Catalog::GetInstance();
+    try {
+      catalog->CreateIndex(database_oid, table_oid, col_oid_vector,
+                           DEFUALT_SCHEMA_NAME, index_name, IndexType::BWTREE,
+                           IndexConstraintType::DEFAULT, is_unique, txn);
+    } catch (CatalogException e) {
+      LOG_ERROR("Create Index Failed");
+      txn_manager.AbortTransaction(txn);
+      return kj::NEVER_DONE;
+    }
+
+    // Index created. Populate it.
+    auto storage_manager = storage::StorageManager::GetInstance();
+    auto table_object =
+        storage_manager->GetTableWithOid(database_oid, table_oid);
+
+    // Create a seq plan to retrieve data
+    std::unique_ptr<planner::SeqScanPlan> populate_seq_plan(
+      new planner::SeqScanPlan(table_object, nullptr, col_oid_vector, false));
+
+    // Create a index plan
+    std::shared_ptr<planner::AbstractPlan> populate_index_plan(
+      new planner::PopulateIndexPlan(table_object, col_oid_vector));
+    populate_index_plan->AddChild(std::move(populate_seq_plan));
+
+    std::vector<type::Value> params;
+    std::vector<ResultValue> result;
+    std::atomic_int counter;
+    std::vector<int> result_format;
+
+    auto callback = [](void *arg) {
+      std::atomic_int *count = static_cast<std::atomic_int *>(arg);
+      count->store(0);
+    };
+
+    // Set the callback and context state.
+    auto &traffic_cop = tcop::TrafficCop::GetInstance();
```

Comment:
Singleton of traffic_cop is used in the test only and should not be in the engine's path. Actually, we should remove those from tests too.

-- Tianyu, Justin & Tianyi